daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
 	dp := &DriverPool{
 		ScalingPolicy: DefaultScalingPolicy(),
 		Logger:        logrus.New(),
-
-		factory: factory,
-		close:   make(chan struct{}),
-		queue:   newDriverQueue(),
+		factory:       factory,
 	}
-	dp.drivers.byID = make(map[string]Driver)
 	return dp
 }
 
 // Start stats the driver pool.
 func (dp *DriverPool) Start(ctx context.Context) error {
-	if dp.running {
+	if dp.stop != nil {
 		return ErrPoolRunning.New()
-	} else if dp.closed {
+	}
+	select {
+	case <-dp.stopped:
 		return ErrPoolClosed.New()
+	default:
+	}
+
+	dp.stop = make(chan struct{})
+	dp.stopped = make(chan struct{})
+	dp.spawn = make(chan struct{})
+	dp.spawnErr = make(chan error)
+	dp.rescale = make(chan struct{}, 1)
+	dp.get = make(chan driverRequest)
+	dp.put = make(chan Driver)
+	dp.drivers.idle = make(map[Driver]struct{})
+	dp.drivers.all = make(map[Driver]struct{})
+
+	dp.target.Set(1)
+
+	dp.wg.Add(3)
+	go dp.runSpawn()
+	go dp.runPolicy()
+	go dp.manageDrivers()
+
+	// wait for a single instance to come up
+	d, err := dp.getDriver(ctx)
+	if err != nil {
+		_ = dp.Stop()
+		return err
 	}
-	target := dp.ScalingPolicy.Scale(0, 0)
-	if err := dp.setInstances(ctx, target); err != nil {
-		_ = dp.setInstances(context.Background(), 0)
+	if err := dp.putDriver(d); err != nil {
 		return err
 	}
-	dp.running = true
-
-	go dp.scaling()
 	return nil
 }
 
-// setInstances changes the number of running driver instances. Instances
-// will be started or stopped as necessary to satisfy the new instance count.
-// It blocks until the all required instances are started or stopped.
-func (dp *DriverPool) setInstances(ctx context.Context, target int) error {
-	if target < 0 {
-		return ErrNegativeInstances.New()
+// runPolicy goroutine re-evaluates the scaling policy on a regular time interval and sets
+// a target number of instances. The scaling itself will be performed by the manager goroutine.
+func (dp *DriverPool) runPolicy() {
+	defer dp.wg.Done()
+	ticker := time.NewTicker(time.Millisecond * 100)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-dp.stop:
+			return
+		case <-ticker.C:
+		}
+		total := dp.running.Value()
+		load := dp.requests.Value()
+		dp.drivers.RLock()
+		idle := len(dp.drivers.idle)
+		dp.drivers.RUnlock()
+
+		target := dp.ScalingPolicy.Scale(total, load-idle)
+		if target < 1 {
+			target = 1 // there should be always at least 1 instance
+		}
+		old := dp.target.Set(target)
+		if old != target {
+			// optionally signal to the manager goroutine
```

Comment:
The `rescale` channel is buffered, thus this will just set a "flag" for the manager goroutine to re-evaluate scaling conditions later.

The benefits that if the pool is starving, and the policy doesn't allow to scale up, the manager goroutine will be blocked waiting for drivers to be put back to the pool. But the scaling policy is not static, it may allow more instances once more requests are waiting for the driver. In this case, this send will unblock the manager goroutine.

The previous implementation I did a few days ago inlined the policy re-evaluation here. But it turned out to be too complex. The policy expects some stable "tick" for re-evaluation, so it's easier to handle it in a separate goroutine.

Reply:
I agree it doesn't make sense to replicate policy evaluation here. My comment was meant more to determine whether we could just get rid of this extra "attempt" to flag the conditions: Since the manager goroutine already has a retry clock, and will retry anyway, it seems to me whether we flag it here or wait for the next tick can't materially change the response the manager has to a starved pool, right?

Reply:
sounds like reasonable simplification suggestion